### ISSUE TITLE
build fails on certain linux distros where /bin/sh is dash

### ISSUE THREAD

**Issue by kaveman-** [rumpkernel/rumprun#38]
build fails on VoidLinux with the following error:

```
nbmake[2]: nbmake[2]: don't know how to make pci.c. Stop
```

**Comment by kaveman-** [comment]
on VoidLinux dash version is 0.5.8_3 and on debian testing (where there is no build failure) dash is 0.5.7-4+b1.

**Comment by anttikantee** [comment]
Is it possible to install dash 0.5.8_3 on Debian just to verify that it's a problem in dash?

**Comment by anttikantee** [comment]
The real problem appears to be that .export statements in the Makefiles get lost between a make and a submake.

**Comment by paulbarbu** [comment]
Did you guys find any solution for this?

My rumprun repo is at version: `dc10b4300af949`
And the submodules:
```
$ git sm status
 5134728efc95d373ecb5b3f75f30791cd655daf2 buildrump.sh (v20140526-462-g5134728)
 8a5d1679075f3a304d609323c3d27e33c4400f47 src-netbsd (remotes/origin/appstack-src)

```
I'm on Debian and I tried starting `build-rr.sh hw` with fish, bash and sh, same result for each of them.

Before updating it worked fine, unfortunately I no longer know the previous revisions.

**Comment by mato** [comment]
Data point: I do builds on Debian (jessie) all the time, my `/bin/sh` is `dash` and I've never seen this problem. I'd be happy to help fix it, but since I can't reproduce it that's kind of hard.

**Comment by anttikantee** [comment]
On 17/02/16 11:07, Martin Lucina wrote:
> Data point: I do builds on Debian (jessie) all the time, my `/bin/sh` is `dash` and I've never seen this problem. I'd be happy to help fix it, but since I can't reproduce it that's kind of hard.

What is says in the audit log is all I know: "on VoidLinux dash version 
is 0.5.8_3 and on debian testing (where there is no build failure) dash 
is 0.5.7-4+b1."

**Comment by paulbarbu** [comment]
My dash version is: 0.5.8-2.1

**Comment by paulbarbu** [comment]
Indeed, on Jessie with dash 0.5.7-4+b1 the build goes OK. Apparently there's a bug in dash? But some time ago it compiled successfully in fish 2.2.0 for me, now with dash 0.5.8 and fish 2.2.0 it doesn't work.

**Comment by mato** [comment]
I've managed to reproduce this with Debian testing (stretch), which has dash 0.5.8-2.1. Will investigate.

**Comment by mato** [comment]
Based on Antti's hint that the problem is in .export statements getting lost I've tracked down the commit that introduces the problem in dash: `46d3c1a` (http://git.kernel.org/cgit/utils/dash/dash.git/commit/?id=46d3c1a614f11f0d40a7e73376359618ff07abcd). Reverting this commit allows the build to proceed.
Am going to rebuild dash with some printf()s to see what variable name(s) it considers "invalid" and is dropping on the floor.

**Comment by mato** [comment]
So, the culprit are exported variables with names containing `.`:
`>>>> not setting invalid name RUMPCOMP_USER_SRCS.rumpdev_pci=rumppci.c rumpdma.c`
Not sure what the correct solution is, I'll report this upstream and follow up.

**Comment by mato** [comment]
From Std 1003.1-2001:
> Environment variable names used by the utilities in the Shell and Utilities volume of IEEE Std  1003.1-2001 consist solely of uppercase letters, digits, and the '_' (underscore) from the characters defined in Portable Character Set and do not begin with a digit. Other characters may be permitted by an implementation; applications shall tolerate the presence of such names.
So, dash is being strictly POSIX compliant and is technically doing the right thing.
@anttikantee What is the right solution on our side? Can we patch `nbmake` to not use `.`?

**Comment by anttikantee** [comment]
Ah, the good old "let's break things because we can" approach.

`.` is just a convention, nothing to do with nbmake per se.  I think it can be replaced by `_` and let caps denote the delimiter between the "subclass" of the variable.  Do you want to prepare/test a patch?

**Comment by mato** [comment]
I've sent a bug report to the dash mailing list (Cc:ed to rumpkernel-users) asking if they'd be willing to relax the behaviour, since there are other users of BSD make besides us. Let's see what they say. If not, then I'll try patching nbmake.

**Comment by anttikantee** [comment]
I don't think changing dash back is useful, since there's nothing to say that other shells might not tighten their conventions to be "posixly correct" in the future.

**Comment by mato** [comment]
Good point. I'll take a look at patching nbmake.

**Comment by anttikantee** [comment]
No, nbmake doesn't need to be touched, just the Makefiles.

**Comment by mato** [comment]
At least `.MAKE.LEVEL` is used internally by nbmake to keep track of the recursion level, so it's not just the Makefiles.

**Comment by anttikantee** [comment]
```
#ifndef MAKE_LEVEL_ENV
# define MAKE_LEVEL_ENV "MAKELEVEL"
#endif
```

Looks good to me.

**Comment by mato** [comment]
For the record, patches to fix this problem are available on the mailing list, here: https://www.freelists.org/post/rumpkernel-users/PATCH-pci-Replace-with--in-exported-var-names,4
Once these are merged I will close this issue.
